src/data/set/basic.lean

```diff
@@ -871,6 +871,15 @@ theorem eq_preimage_subtype_val_iff {p : α → Prop} {s : set (subtype p)} {t :
 ⟨assume s_eq x h, by rw [s_eq]; simp,
  assume h, ext $ assume ⟨x, hx⟩, by simp [h]⟩
 
+lemma if_preimage (p : α → Prop) [decidable_pred p] (f g : α → β) (s : set β) :
+  (λa, if p a then f a else g a)⁻¹' s = (p ∩ f ⁻¹' s) ∪ (-p ∩ g ⁻¹' s) :=
```

Comment:
This will result in an ugly RHS if `p` is not a `λ x, x ∈ t` for some `t : set α`.